tmt/package_managers/apk.py

```diff
@@ -0,0 +1,164 @@
+    Installable,
+    Options,
+    Package,
+    PackagePath,
+    escape_installables,
+    provides_package_manager,
+    )
+from tmt.utils import (
+    Command,
+    CommandOutput,
+    GeneralError,
+    RunError,
+    ShellScript,
+    )
+
+ReducedPackages = list[Union[Package, PackagePath]]
+
+PACKAGE_PATH = {
+    '/usr/bin/flock': 'flock'
+    }
+
+
+@provides_package_manager('apk')
+class Apk(tmt.package_managers.PackageManager):
+    probe_command = Command('apk', '--version')
+
+    install_command = Command('add')
+
+    _sudo_prefix: Command
+
+    def prepare_command(self) -> tuple[Command, Command]:
+        """ Prepare installation command for apk """
+
+        if self.guest.facts.is_superuser is False:
+            self._sudo_prefix = Command('sudo')
+
+        else:
+            self._sudo_prefix = Command()
+
+        command = Command()
+
+        command += self._sudo_prefix
+        command += Command('apk')
+
+        return (command, Command())
+
+    def path_to_package(self, path: FileSystemPath) -> Package:
+        """
+        Find a package providing given filesystem path.
+
+        This is not easily possible in Alpine. There is `apk-file` utility
+        available but it seems unrealiable. Support only a fixed set
+        of mappings until a better solution is available.
+        """
+
+        string_path = str(path)
+
+        if string_path in PACKAGE_PATH:
+            return Package(PACKAGE_PATH[string_path])
+
+        raise GeneralError(f"Unsupported package path '{path} for Alpine Linux.")
+
+    def _reduce_to_packages(self, *installables: Installable) -> ReducedPackages:
+        packages: ReducedPackages = []
+
+        for installable in installables:
+            if isinstance(installable, (Package, PackagePath)):
+                packages.append(installable)
+
+            elif isinstance(installable, FileSystemPath):
+                packages.append(self.path_to_package(installable))
+
+            else:
+                raise GeneralError(f"Package specification '{installable}' is not supported.")
+
+        return packages
+
+    def _construct_presence_script(
+            self, *installables: Installable) -> tuple[ReducedPackages, ShellScript]:
+        reduced_packages = self._reduce_to_packages(*installables)
+
+        shell_script = ShellScript(
+            f'apk info -e {" ".join(escape_installables(*reduced_packages))}')
+
+        return reduced_packages, shell_script
+
+    def check_presence(self, *installables: Installable) -> dict[Installable, bool]:
+        reduced_packages, presence_script = self._construct_presence_script(*installables)
+
+        try:
+            output = self.guest.execute(presence_script)
+            stdout, stderr = output.stdout, output.stderr
+
+        except RunError as exc:
+            stdout, stderr = exc.stdout, exc.stderr
+
+        if stdout is None or stderr is None:
+            raise GeneralError("apk presence check output is None")
```

Comment:
Maybe rather "empty" or "unset" or "unknown" than "None". To a user, "None" means nothing.

Reply:
ack

Reply:
addressed